sdks/python/apache_beam/coders/coders.py

```diff
@@ -1383,22 +1375,74 @@ def from_runner_api_parameter(payload, components, context):
         write_state_threshold=int(payload))
 
 
-class RunnerAPICoderHolder(Coder):
+class ElementTypeHolder(typehints.TypeConstraint):
+  """A dummy element type for external coders that cannot be parsed in Python"""
+  def __init__(self, coder, context):
+    self.coder = coder
+    self.context = context
+
+
+class ExternalCoder(Coder):
   """A `Coder` that holds a runner API `Coder` proto.
 
   This is used for coders for which corresponding objects cannot be
   initialized in Python SDK. For example, coders for remote SDKs that may
   be available in Python SDK transform graph when expanding a cross-language
   transform.
   """
-  def __init__(self, proto):
-    self._proto = proto
 
-  def proto(self):
-    return self._proto
+  coder_count = 0
 
-  def to_runner_api(self, context):
-    return self._proto
+  def __init__(self, element_type_holder):
+    self.element_type_holder = element_type_holder
 
-  def to_type_hint(self):
-    return Any
+  def as_cloud_object(self, coders_context=None):
+    if not coders_context:
+      raise Exception(
+          'coders_context must be specified to correctly encode external coders'
+      )
+    coder_id = coders_context.get_by_proto(
+        self.element_type_holder.coder, deduplicate=True)
+
+    coder_proto = self.element_type_holder.coder
+
+    kind_str = 'kind:external' + str(ExternalCoder.coder_count)
+    ExternalCoder.coder_count = ExternalCoder.coder_count + 1
+    component_encodings = []
+    if coder_proto.spec.urn == 'beam:coder:kv:v1':
```

Comment:
What about correctly handling `kind:stream`? Perhaps others? It seems it would be preferable to instead try to only use external_coder for the unknown leafs rather than duplicate this special-dataflow logic here.